### PR TITLE
SDSTOR-11409 : Fix sanitization issue in fiber btree UT

### PR DESCRIPTION
**Pull request by shosseinimotlagh**

### CODE REVIEW

File: src/tests/test_mem_btree.cpp [hunk 1]
```diff
@@ -77,6 +77,8 @@ struct BtreeTest : public testing::Test {
     using K = typename TestType::KeyType;
     using V = typename TestType::ValueType;
 
+    virtual ~BtreeTest() override = default;
```
Comment (szmyd) [inline src/tests/test_mem_btree.cpp:80]: Just `override`; we want there to be an error if the underlying removes the virtual attribute.

File: src/include/homestore/btree/mem_btree.hpp [hunk 2]
```diff
@@ -43,7 +43,7 @@ class MemBtree : public Btree< K, V > {
     btree_status_t write_node_impl(const BtreeNodePtr& node, void* context) { return btree_status_t::success; }
 
     btree_status_t read_node_impl(bnodeid_t id, BtreeNodePtr& node) const override {
-        node = BtreeNodePtr{r_cast< BtreeNode* >(id)};
+        node.reset(r_cast< BtreeNode* >(id));
```
Comment (szmyd) [inline src/include/homestore/btree/mem_btree.hpp:46]: It's not clear to me what issue this change fixed?

File: src/include/homestore/btree/detail/btree_node.hpp [hunk 3]
```diff
@@ -619,8 +619,10 @@ class BtreeNode : public sisl::ObjLifeCounter< BtreeNode > {
 
     friend void intrusive_ptr_release(BtreeNode* node) {
         if (node->m_refcount.decrement_testz(1)) {
+            auto node_buffer = node->m_phys_node_buf;
             node->~BtreeNode();
             delete[] uintptr_cast(node);
+            delete[] node_buffer;
```
Comment (szmyd) [inline src/include/homestore/btree/detail/btree_node.hpp:625]: It seems odd to deallocate in a different order than allocation. Originally the BtreeNode is allocated, then the node_buf is allocated, constructed and finally the BtreeNode is constructed. Should we not to the same in reverse? :
```
            auto node_buffer = node->m_phys_node_buf;
            node->~BtreeNode();
            delete[] node_buffer;
            delete[] uintptr_cast(node);
```

Comment (szmyd) [reply src/include/homestore/btree/detail/btree_node.hpp:625]: Also, it's not clear that `node_buffer` was allocated with `new[]` anywhere. We should define a specific type for this so someone cannot allocate with `new` or `malloc` and expect this class to work right?

File: src/include/homestore/btree/mem_btree.hpp (hunk 2 above)
Comment (shosseinimotlagh) [reply src/include/homestore/btree/mem_btree.hpp:46]: the way that that has been implemented it causes the reference count for the intrusive pointer to bump up. As a result, in [post_order_traversal](https://raw.githubusercontent.com/eBay/HomeStore/fiber/src/include/homestore/btree/detail/btree_common.ipp#:~:text=read_and_lock_node(m_root_node_info.bnode_id()%2C%20root%2C%20ltype%2C%20ltype%2C%20nullptr)%3B), BtreeNodePtr root has not release the pointer.

File: src/tests/test_mem_btree.cpp (hunk 1 above)
Comment (shosseinimotlagh) [reply src/tests/test_mem_btree.cpp]: I removed it. I did as you suggested, destructor for each test would be needed( which is unnecessary).

File: src/include/homestore/btree/mem_btree.hpp (hunk 2 above)
Comment (szmyd) [reply src/include/homestore/btree/mem_btree.hpp:46]: Thanks for the explanation!

File: src/include/homestore/btree/detail/btree_node.hpp (hunk 3 above)
Comment (shosseinimotlagh) [reply src/include/homestore/btree/detail/btree_node.hpp:625]: The sequence is 
create_root_node() -> alloc_leaf_node() -> alloc_node() [[node_buffer is created here]](https://raw.githubusercontent.com/eBay/HomeStore/fiber/src/include/homestore/btree/mem_btree.hpp#:~:text=BtreeNodePtr%20alloc_node(bool%20is_leaf)%20override%20%7B), then new_node is created and the allocated [node_buffer is passed](https://raw.githubusercontent.com/eBay/HomeStore/fiber/src/include/homestore/btree/detail/btree_node_mgr.ipp#:~:text=static%20BtreeNode*%20create_node(uint32_t%20node_ctx_size%2C%20Args%26%26...%20args)%20%7B) .

**Review by szmyd — APPROVED** [review]

**Pull request merged**